docs/2019-12-16-Question-Mark-Variables-With-Ensures.md

```diff
@@ -0,0 +1,272 @@
+`fresh(Is)` _transitions_ to an integer not in `Is` (could be a different one
+each time it is used), while `arbDifferentFromList(Is)` _is equal to_ a (fixed)
+integer not in `Is`.
+
+### Skolemization
+
+One possible approach to implementing the abovewould be through [Skolemization](https://en.wikipedia.org/wiki/Skolem_normal_form),
+i.e., by replacing the existential variables with fresh uninterpreted functions
+applied to the universally quantified variables.
+
+For example, the `(ARBDIFF)` rule above could be Skolemized to
+
+```
+(ARBDIFF') \forall Is:List{Is} . \neg(arbDiffI(Is) \in Is) /\ arbDifferentFromList(Is) = arbDiffI(Is)
+```
+where
+```
+arbDiffI : List{Int} -> Int
+```
+
+is a symbol satisfying the `(Function)` axiom, without further constraints.
+
+### Frontend Syntax
+
+We do not need special frontend syntax to define `arb()`.
+We only need to define it in the usual way as a function
+(instead of a language construct), and provide no axioms for it
+(the `(Function)` axiom is generated automatically from the `function`
+attribute annotation.
+
+The frontend doesn't currently provide a syntax to define the variants of
+`arb()`, such as `arbDifferentFromList(Is)` in the above.
+
+We envision two possible solutions:
+
+1. they are not allowed, since the users could themselves Skolemize the axiom
+
+    ```
+    syntax Int ::= arbDiffI(List{Int}) [function, functional]
+    syntax Int ::= arbDifferentFromList(List{Int}) [function, functional]
+    rule arbDifferentFromList(Is:List{Int}) => arbDiffI(Is)
+      ensures notBool (arbDiffI(Is) inList{Int} Is)
+    ```
+
+2. They are allowed, assuming that a `function` annotation in the syntax
+   definition, together with the use of `?` variables in the `rhs` of the rule and
+   the `ensures` clause would trigger Skolemization and the generation of
+   the rule above.
+   That is, `adbDifferentFromList` could be written in K as
+
+    ```
+    syntax Int ::= arbDifferentFormList(List{Int}) [function]
+    rule arbDifferentFromList(Is:List{Int}) => ?I:Int
+      ensures notBool (?I inList{Int} Is)
+    ```
+
+## Example D: Interval (Non-function Symbols) `interval()`
+
+The symbol `interval(M,N)` is not a PL construct, nor a function in the
+first-order sense, but a proper matching-logic symbol, whose interpretation is
+in the powerset of its domain.
+Its axioms will not use rewrites but equalities.
+
+### Axioms
+
+The intended semantics of `interval(M,N)` is that it equals the _set_ of
+integers that are larger than or equal to `M` and smaller than or equal to `N`.
+Its axiom is
+
+```
+(INTERVAL) \forall M:Int . \forall N:Int . interval(M,N) = \exists X:Int . X:Int /\ X >= M /\ X <= N
+```
+
+### Frontend Syntax
+
+The symbol declaration would require a special attribute to signal the fact that
+it represent a proper ML symbol which is defined equationally. The precise
+name would have to be establishes, but let's call it here `power-function`
```

Comment:
estabilished

Reply:
`power-function` ... hm, then all ML symbols are power-functions by default.  Since almost all the ML symbols in a definition end up being function symbols or constructors, the poor proper ML symbols lost their default status.  Not sure what name is better here: power-function, relation, proper-symbol, bare-symbol, default-symbol, etc.

Reply:
my option for power-function was to signal that its definition is to be done through equations, since the front-end only uses the `=>` symbol for both rewrites and equalities.